### ISSUE TITLE
[patch] Greatly improve telnet performance

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2107]
|<img alt="shieldszero's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| shieldszero reported|
|-|-|
|Trac ID|trac#2107|
|Type|enhancement|
|Created|2006-09-18 20:15:19Z|
|Branch|https://github.com/twisted/twisted/tree/optimise-telnet-2107|

The current version of twisted.conch.telnet.Telnet.dataReceived() calls applicationDataReceived() for _every character_, which is correctly described in a comment as "grossly inefficient".  This patch corrects that and makes programs that process significant amounts of telnet data many times faster.

Attachments:

* [telnet-performance.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/fe8/fe8ed5b1a81d6ba010e4dbd544dd0ef199cd1bb8/f91e48e4e728c4075b129a84335e05e9cf5b3332.diff) (2970 bytes) - added by shieldszero on 2006-09-18 20:15:31Z - 
* [telnet-performance.2.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/fe8/fe8ed5b1a81d6ba010e4dbd544dd0ef199cd1bb8/457fd06d151c32e7b785bf4db7dc8011384c9da3.diff) (2966 bytes) - added by shieldszero on 2006-09-18 20:51:48Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__2107 2107
type__enhancement enhancement
reporter__shieldszero shieldszero
priority__highest highest
milestone__ 
branch__branches_optimise_telnet_2107 branches/optimise-telnet-2107
branch_author__ 
status__closed closed
resolution__fixed fixed
component__conch conch
keywords__ 
time__1158610519000000 1158610519000000
changetime__1200871989000000 1200871989000000
version__None None
owner__ 
cc__shieldszero cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

```
437	437	    def dataReceived(self, data): 
438	 	        # Most grossly inefficient implementation ever 
 	438	 
 	439	        # Create a buffer to avoid calling applicationDataReceived() for 
 	440	        # every character, which is grossly inefficient. 
 	441	        buf = [] 
 	442	        def flush_buf(buf=buf): 
 	443	            if buf: 
 	444	                self.applicationDataReceived(''.join(buf)) 
 	445	                buf = [] 
 	446	 
439	447	        for b in data: 
```

The buf assignment above is a no-op.  Something like ```buf[:] = []``` would work, though.

**Comment by twisted-trac** [comment]
|[<img alt="oubiwann's avatar" src="https://avatars.githubusercontent.com/u/894251?s=50" width="50" height="50">](https://github.com/oubiwann)<a name="note_2"></a>|@oubiwann commented|
|-|-|

shieldszero,

You need to submit patches in the form described here:

  [TwistedDevelopment](https://twistedmatrix.com/trac/wiki/TwistedDevelopment)

(```diff -u```)

**Comment by twisted-trac** [comment]
|<img alt="shieldszero's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|shieldszero commented|
|-|-|

Replying to [oubiwann](#note_2):
> You need to submit patches in the form described here:
> 
>   [TwistedDevelopment](https://twistedmatrix.com/trac/wiki/TwistedDevelopment)
> 
> (```diff -u```)

This patch is already in ```diff -u``` format.

**Comment by twisted-trac** [comment]
|<img alt="shieldszero's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|shieldszero commented|
|-|-|

Replying to [exarkun](#note_1):
> The buf assignment above is a no-op.  Something like ```buf[:] = []``` would work, though.

Thanks, good catch.  Here is a corrected patch, telnet-performance.2.diff.

**Comment by twisted-trac** [comment]
|[<img alt="oubiwann's avatar" src="https://avatars.githubusercontent.com/u/894251?s=50" width="50" height="50">](https://github.com/oubiwann)<a name="note_6"></a>|@oubiwann commented|
|-|-|

Replying to [shieldszero](#note_3):
> > You need to submit patches in 
> > 
> > (```diff -u```)
> 
> This patch is already in ```diff -u``` format.

shieldszero, my bad; I totally didn't look at the patch file, assming incorrectly based on JP's copy and paste.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

It would be good to have a test for this.  For example, I think it would be sensible to assert that applicationDataReceived is called with a string of a particular length, given a particular input, rather than with one or two bytes only.

At the moment, I feel like I will probably want to do this soon, perhaps after [#1967](https://github.com/twisted/twisted/issues/1967) is merged (so that there is a convenient place to put the tests - on TelnetTests).  Time permitting, of course.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

(In [22374]) Branching to 'optimise-telnet-2107'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun removed owner|
|-|-|

Wrote a lot of tests for different codepaths through Telnet.dataReceived and basically re-implemented shieldzero's patch to make them path.  I didn't end up with quite as many buffer flushes as him, which I think is fine.  I also didn't wrap up buffer flushing in a function, since the point is to make it faster.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_11"></a>|@therve set owner to @exarkun|
|-|-|

Nothing to say, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun set status to `closed`|
|-|-|

(In [22380]) Merge optimise-telnet-2107

Author: exarkun
Reviewer: therve
Fixes [#2107](https://github.com/twisted/twisted/issues/2107)

Reduce the number of calls into application code to deliver application-level
data in twisted.conch.telnet.Telnet.dataReceived by buffering application data
as long as is possible.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
